### PR TITLE
feat(download): Change .zip library to allow download backups greater…

### PR DESCRIPTION
**Pull request by egalvis27**
## What is Changed / Added
----
- Reworked backup zip generation to a streaming approach in zip service, using Archiver for large archives.
- Added new zip flow wiring in network layer and facade to support large backup downloads end to end.
- Refactored download logic into small focused modules under network/download and network/download/stream.
- Split key steps into dedicated units: metadata resolution, key resolution, stream decryption, progress stream, temp file write, retry attempt, and v2-to-v1 fallback.
- Migrated filesystem usage in download path to promise-based APIs where needed.
- Added broad unit tests for new download and stream modules, plus zip service coverage.
- Fixed stream typings in tests by using node stream/web ReadableStream and WritableStream types.

## Why
- Old zip path could fail or become unstable with very large backups over 4GB.
- Streaming zip reduces memory pressure and handles big backups safely.
- Smaller modules make code easier to read, maintain, and test.
- Explicit fallback and error mapping improve reliability across file versions and edge cases.
- More tests reduce regression risk for backup download behavior.

### CODE REVIEW

File: src/backend/features/download/stream/build-progress-stream.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { ReadableStream } from 'node:stream/web';
+
+type Props = {
+  source: ReadableStream<Uint8Array>;
+  onRead: (readBytes: number) => void;
+};
+
+export function buildProgressStream({ source, onRead }: Props): ReadableStream<Uint8Array> {
```
Comment (AlexisMora) [inline src/backend/features/download/stream/build-progress-stream.ts:8]: What is the purpose of this method?

File: src/backend/features/download/stream/convert-to-writable-stream.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+export function convertToWritableStream({ writeStream }: Props): WritableStream<Uint8Array> {
+  let isClosed = false;
+
+  return new WritableStream<Uint8Array>({
+    async write(chunk) {
+      if (isClosed || writeStream.writableEnded || writeStream.destroyed) {
+        throw new Error('Write stream already closed');
+      }
+
+      const buffer = Buffer.from(chunk);
```
Comment (AlexisMora) [inline src/backend/features/download/stream/convert-to-writable-stream.ts:15]: So, essentially you are writing the chunks into an in memory buffer no?

File: src/backend/features/download/stream/convert-to-writable-stream.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import type { WriteStream } from 'node:fs';
+import { WritableStream } from 'node:stream/web';
+
+type Props = { writeStream: WriteStream };
+
+export function convertToWritableStream({ writeStream }: Props): WritableStream<Uint8Array> {
```
Comment (AlexisMora) [inline src/backend/features/download/stream/convert-to-writable-stream.ts:6]: what is this convert really doing? you are already passing a writable, no?

File: src/backend/features/download/stream/create-empty-readable-stream.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { ReadableStream } from 'node:stream/web';
+
+export function createEmptyReadableStream() {
+  return new ReadableStream<Uint8Array>({
```
Comment (AlexisMora) [inline src/backend/features/download/stream/create-empty-readable-stream.ts:4]: Why is this necesary?

File: src/backend/features/download/stream/create-empty-readable-stream.test.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { createEmptyReadableStream } from './create-empty-readable-stream';
+
+describe('create-empty-readable-stream', () => {
+  it('should return a stream that ends immediately', async () => {
```
Comment (AlexisMora) [inline src/backend/features/download/stream/create-empty-readable-stream.test.ts:4]: why

File: src/apps/main/network/download/stream/get-decrypted-stream.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
+import type { Decipheriv } from 'node:crypto';
+import { ReadableStream } from 'node:stream/web';
+
+type JoinPops = { streams: ReadableStream<Uint8Array>[] };
```
Comment (AlexisMora) [inline src/apps/main/network/download/stream/get-decrypted-stream.ts]: is this on purpose or there is a typo and should be "JoinProps"?

File: src/apps/main/network/download/stream/get-decrypted-stream.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+function joinReadableBinaryStreams({ streams }: JoinPops): ReadableStream {
```
Comment (AlexisMora) [inline src/apps/main/network/download/stream/get-decrypted-stream.ts]: If this should not beexported then you could add it at the end of the file, furthermore, you could even make tests for this

File: src/backend/features/download/stream/index.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export { buildProgressStream } from './build-progress-stream';
+export { convertToWritableStream } from './convert-to-writable-stream';
+export { createEmptyReadableStream } from './create-empty-readable-stream';
+export { getDecryptedStream } from './get-decrypted-stream';
+export { isEmptyBackupFileWithoutFileId } from './is-empty-backup-file-without-file-id';
+export { writeDownloadStreamToFile } from './write-download-stream-to-file';
```
Comment (AlexisMora) [inline src/backend/features/download/stream/index.ts:6]: I have checked that `writeDownloadStreamToFile`, `isEmptyBackupFileWithoutFileId `, `createEmptyReadableStream`, `convertToWritableStream ` is not being used outside the download/stream module, there is no need to export them right? also, why did you chose to create the module here? thought that we agreed to follow the backend/features architecture

File: src/backend/features/download/stream/write-download-stream-to-file.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  tempFilePath: string;
+  abortController?: AbortController;
+};
+
+export async function writeDownloadStreamToFile({ stream, tempFilePath, abortController }: Props) {
+  const fileHandle = await open(tempFilePath, 'w');
+  const writeStream = fileHandle.createWriteStream();
+
+  try {
+    await stream.pipeTo(convertToWritableStream({ writeStream }), {
```
Comment (AlexisMora) [inline src/backend/features/download/stream/write-download-stream-to-file.ts:19]: why are you convertiing the WriteStream to a WritableStream? is it really necessary?

File: src/backend/features/download/stream/build-progress-stream.ts (hunk 1 above)
Comment (egalvis27) [reply src/backend/features/download/stream/build-progress-stream.ts:8]: A wrapper that implements a stream to track the number of bytes read in real time without modifying the data stream. It allows the progress bar to be updated in real time during downloads.

File: src/backend/features/download/stream/convert-to-writable-stream.ts (hunk 3 above)
Comment (egalvis27) [reply src/backend/features/download/stream/convert-to-writable-stream.ts:6]: This function acts as a bridge between the download stream, which uses an older Node.js interface, and the stream that is needed to write the ZIP file, which uses a more up-to-date method., allowing the downloaded data to be written to the ZIP file. This functionality already existed before, but these functions were all in the same file.

File: src/backend/features/download/stream/convert-to-writable-stream.ts (hunk 2 above)
Comment (egalvis27) [reply src/backend/features/download/stream/convert-to-writable-stream.ts:15]: If it's exactly as you describe, the download creates a Node.js stream that is converted into a WebStream so it can be saved to the ZIP file.

File: src/backend/features/download/stream/create-empty-readable-stream.ts (hunk 4 above)
Comment (egalvis27) [reply src/backend/features/download/stream/create-empty-readable-stream.ts:4]: This feature is used to create an empty file inside the backup ZIP file in case there is already one in the backup being downloaded, in order to prevent errors or customer complaints that the backup is incomplete.

File: src/backend/features/download/stream/create-empty-readable-stream.test.ts (hunk 5 above)
Comment (egalvis27) [reply src/backend/features/download/stream/create-empty-readable-stream.test.ts:4]: Since it's empty, it's something that gets done right away

File: src/apps/main/network/download/stream/get-decrypted-stream.ts (hunk 6 above)
Comment (egalvis27) [reply src/apps/main/network/download/stream/get-decrypted-stream.ts]: That's a mistake; I forgot to add the “r.”

File: src/apps/main/network/download/stream/get-decrypted-stream.ts (hunk 7 above)
Comment (egalvis27) [reply src/apps/main/network/download/stream/get-decrypted-stream.ts]: I moved it and added a test

File: src/backend/features/download/stream/index.ts (hunk 8 above)
Comment (egalvis27) [reply src/backend/features/download/stream/index.ts:6]: I created it here so as not to move the code around too much, but now that you mention it, we could move the entire “download” folder and still have the same number of changes.

File: src/backend/features/download/stream/write-download-stream-to-file.ts (hunk 9 above)
Comment (egalvis27) [reply src/backend/features/download/stream/write-download-stream-to-file.ts:19]: The stream that `stream.pipeTo` expects is different from the one we have available. `stream.pipeTo` uses a new interface, but what is received when it is downloaded is an old interface; the conversion is used to avoid incompatibility between the two streams.

File: src/core/bootstrap/register-process-handlers.ts [hunk 10]
```diff
@@ -1,5 +1,13 @@
 import { logger } from '@internxt/drive-desktop-core/build/backend';
 
+function isContentLengthMismatch(error: unknown) {
+  if (!(error instanceof Error)) {
+    return false;
+  }
```
Comment (egalvis27) [inline src/core/bootstrap/register-process-handlers.ts:6]: Explicit handling of this error has been added because, during the download process, it is possible that fewer bytes than requested may be received or that the content-length header may be incorrect. Explicit handling of this error prevents the download from stopping; this error is temporary, and any download that has not been completed is retried.

**Review by AlexisMora — APPROVED** [review]
